### PR TITLE
sp-{io,core}: replace secp256k1/libsecp256k1 with k256

### PR DESCRIPTION
**Pull request by Gioyik**
Follow-up to #5841. Moves the last users of the bitcoin-core `secp256k1` bindings and the pure-Rust `libsecp256k1` crate to `k256`: the `secp256k1_ecdsa_recover{_compressed}` host functions (both versions) and `sp_core::ecdsa::Pair::verify_deprecated`. Both crates remain only as dev-dependencies, serving as oracles for differential tests.

The host function ABI is frozen: old runtime blobs call these functions during block re-execution, so results must match bit-for-bit. Preserved per version:

- v1: overflowing parsing (`r`/`s` >= n reduced modulo n, `BadRS` unreachable, zero fails recovery with `BadSignature`).
- v2: strict parsing (`r`/`s` >= n gives `BadRS`, zero parses and fails recovery with `BadSignature`).
- Both: high-S signatures stay accepted. k256 rejects them at recovery, so they are normalized with the recovery-id parity flipped, as #5841 did in `sp-core`.
- `sp-io` loses its implicit `secp256k1` cargo feature (major bump).
- k256 recovery is slower than the C library; this is what I read but haven't checked nor tested the impact of it, probably required re-run of benchmarks (?).

Differential tests confirm that the boundary values (zero, n and neighbours, the n/2 low/high-S boundary, p-n and neighbour, the maximum, pseudo-random scalars, valid and invalid recovery bytes) and assert SCALE-identical results against verbatim copies of the historical implementations.

### CODE REVIEW

File: substrate/primitives/io/Cargo.toml [hunk 1]
```diff
@@ -51,16 +49,23 @@ polkavm-derive = { workspace = true }
 [build-dependencies]
 rustversion = { workspace = true }
 
+[dev-dependencies]
+# Only used to check that the k256-based implementations of the frozen
+# `secp256k1_ecdsa_recover*` host functions behave bit-for-bit like the historical
+# backends they replaced.
```
Comment (bkchr) [inline substrate/primitives/io/Cargo.toml:55]:
```suggestion
```

File: substrate/primitives/core/Cargo.toml [hunk 2]
```diff
@@ -76,6 +75,9 @@ dyn-clone = { workspace = true }
 
 [dev-dependencies]
 criterion = { workspace = true, default-features = true }
+# Only used to check that the k256-based `Pair::verify_deprecated` behaves exactly like
+# the historical `libsecp256k1`-based implementation.
```
Comment (bkchr) [inline substrate/primitives/core/Cargo.toml:79]:
```suggestion
```

**Comment by athei** [comment]
I would expect the dependencies to also be removed from the root `Cargo.toml` or do we have some users left?

File: substrate/primitives/core/src/ecdsa.rs [hunk 3]
```diff
@@ -865,47 +862,126 @@ mod test {
...
+					raw[64] = v;
+					let sig = Signature::from_raw(raw);
+
+					let legacy = libsecp_recover_overflowing(&raw, &msg_hash);
+					assert_eq!(
+						Pair::verify_deprecated(&sig, message, &public),
+						legacy == Some(public.0),
+						"implementations diverged for r={r:02x?} s={s:02x?} v={v}",
+					);
+					if let Some(key) = legacy {
```
Comment (bkchr) [inline substrate/primitives/core/src/ecdsa.rs:976]: Maybe we should assert that we at least once recover a public key?

**Review by bkchr — APPROVED** [review]
